others/globals_and_utils.py

```diff
@@ -218,6 +196,118 @@ def load_config(filename: str) -> dict:
+                for (type,path,change) in diff:
+                    if type=='change':
+                        (new,old)=change
+                        log.info(f'{path} changed: old/new {old}/{new}')
+                        if changes is None:
+                            changes=dict()
+                        key=path.split('.')[-1]
+                        changes[key]=new
+
+            load_or_reload_config_if_modified.mtimes[filepath]=mtime
+            load_or_reload_config_if_modified.cached_configs[filepath]=new_config_obj
+            log.info(f'(re)loaded modified config (File "{filepath}")') # format (File "XXX") generates pycharm link to file in console output
+            if not changes is None and not target_obj is None:
+                update_attributes(changes, target_obj)
+            return (new_config_obj,changes) # it was modified, so return changes dict
+        else:
+            return (load_or_reload_config_if_modified.cached_configs[filepath], None) # return previous config and None for changes
+    except Exception as e:
+        logging.error(f'could not reload {filepath}: got exception {e}')
+        raise e
+
+load_or_reload_config_if_modified.cached_configs=dict()
+load_or_reload_config_if_modified.mtimes=dict()
+load_or_reload_config_if_modified.start_time=time.time()
+load_or_reload_config_if_modified.counter=dict()
+
+import numbers
+
+
+def update_attributes(updated_attributes: "dict[str, TensorType]", target_obj):
+    """ Update scalar float32 attributes in compiled code (tensorflow JIT) that have changed, i.e. copy them to the compiled/GPU instance.
+
+    After this call, such attribute values are available to the TF function as self.key, where key is the key used in dict.
+
+    Used in various controllers in Control_Toolkit/Control_Toolkit_ASF_Template/Controllers.
+
+    :param updated_attributes: a dict with string keys and scalar numeric value attribute to set.
+    :param target_obj: the object into which we set the attribute.
+
+    """
+    for property, new_value in updated_attributes.items():
+        if hasattr(target_obj, property) :  # make sure it is mutable if we want to set it
+            objtype=None
+            if isinstance(new_value,numbers.Integral):
+                objtype=target_obj.lib.int32
+            elif isinstance(new_value,numbers.Real):
+                objtype=target_obj.lib.float32
+            elif isinstance(new_value,str):
+                objtype=target_obj.lib.string
+            elif isinstance(new_value,np.ndarray):
+                objtype=target_obj.lib.float32  # todo assuming all np arrays should go to float 32 here
+            else:
+                log.warning(f'attribute "{property}" has unknown object type {type(new_value)}; cannot assign it')
+            if objtype:
+                target_obj.lib.assign(getattr(target_obj, property), target_obj.lib.to_variable(new_value,objtype))
```

Comment:
```suggestion
                p = getattr(target_obj, property)
                if target_obj.lib.dtype(p) != objtype:
                    log.warning(f"New value {new_value} for property {property} in {target_obj} has wrong type. Currently has type {target_obj.lib.dtype(p)}, but {objtype} was provided.")
                target_obj.lib.assign(p, target_obj.lib.cast(new_value,objtype))
```
Reason: Assign operation does not require a variable passed to it, as long as `target_obj.property` is already a variable.